solr/core/src/java/org/apache/solr/core/HttpSolrClientProvider.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.core;
+
+import static org.apache.solr.util.stats.InstrumentedHttpRequestExecutor.KNOWN_METRIC_NAME_STRATEGIES;
+
+import java.lang.invoke.MethodHandles;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import org.apache.solr.client.solrj.SolrClient;
+import org.apache.solr.client.solrj.impl.Http2SolrClient;
+import org.apache.solr.client.solrj.impl.HttpClientUtil;
+import org.apache.solr.common.SolrException;
+import org.apache.solr.common.util.IOUtils;
+import org.apache.solr.metrics.SolrMetricManager;
+import org.apache.solr.metrics.SolrMetricsContext;
+import org.apache.solr.security.HttpClientBuilderPlugin;
+import org.apache.solr.update.UpdateShardHandlerConfig;
+import org.apache.solr.util.stats.HttpClientMetricNameStrategy;
+import org.apache.solr.util.stats.InstrumentedHttpListenerFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Internal use only. Provider of the default SolrClient implementation. */
+public class HttpSolrClientProvider implements SolrInfoBean {
```

Comment:
Why does it implement SolrInfoBean?  It's for mostly "pluggable" things, unlike this.

Reply:
SolrMetricsProducer is probably not needed as we are likely directly/manually initializing producing metrics, so the interface accomplishes nothing.  I think.

Reply:
We have two options for passing the parent context used to initialize the HttpListenerFactory for HttpClient: either pass it through the constructor or create a method. We're using the second approach, where SolrMetricProducer requires subclasses to implement initializeMetrics. This method handles the initialization of InstrumentedHttpListenerFactory and also serves as a marker.

Reply:
It also take cares of closing of solrMatricContext. However, I forgot to call the super close method which deal with it. Will be adding it.

Reply:
Actually I removed the interface. I thought about it and It is only valid If we want to expand the scope further. But for HttpSolrClientProvider that is not needed, So i dropped it.